### PR TITLE
Performance improvements

### PR DESCRIPTION
**Pull request by bcoltin**
Improve threading for reading data, measured about a 50% improvement in benchmarking dataset with worldview images.

Also removed resume mode. With a different randomization of the image ordering each time we run, doesn't matter if we re-run part of an epoch.

### CODE REVIEW

**Comment by ScottMcMichael** [comment]
Tensorflow is supposed to be pipelining input data loading to the point where we should not need to be doing it manually ourselves.  Were you able to get any speed improvements by experimenting with the tensorflow dataset parameters?  It could be the case that our image reading code is just not compatible with that dataset loading functionality.

The resume feature was added to handle the case where we could only process part of the dataset, say 60% of images, before training was interrupted.  Is the new procedure to just pick a number of times to repeat that process (with a random set of images each time)?

**Comment by bcoltin** [comment]
"Supposed to" is the key word... Actually, using multiple threads through tensorflow slows things down. It seems there is a very poor scheduler--- it will not load things in order, at all. So it will be blocked waiting for the next example to load in the GPU, while it loads hundreds of other training examples. :( So my changes are faster than single threaded tensorflow, which is much faster than multi-threaded... I suspect what they're doing works better when each example is faster to load, but ours are bigger than the standard? Not entirely sure.

Yeah, that's right, there's not really much gain from carefully making sure we get every example, to doing a random selection multiple times, for the case where the dataset is very large.

**Comment by ScottMcMichael** [comment]
I never was able to run down why training across multiple GPU nodes on tallgrass slowed down training...

File: delta/imagery/imagery_dataset.py [hunk 1]
```diff
@@ -76,117 +72,11 @@ def __init__(self, images, labels, output_shape, chunk_shape, stride=None,
             assert len(images) == len(labels)
         self._images = images
         self._labels = labels
-        self._access_counts = [np.zeros(0, np.uint8), np.zeros(0, np.uint8)]
+        self._epoch = [0, 0]
```
Comment (ScottMcMichael) [inline delta/imagery/imagery_dataset.py:75]: Add comment why there are two numbers here

File: delta/imagery/imagery_dataset.py [hunk 2]
```diff
@@ -253,42 +121,66 @@ def _tile_generator(self, i, is_labels): # pragma: no cover
...
-        # read one row ahead of what we process now
-        next_buf = self._iopool.submit(lambda: image.read(tiles[0][0]))
-        for (c, (rect, sub_tiles)) in enumerate(tiles):
-            cur_buf = next_buf
-            if c + 1 < len(tiles):
-                # extra lambda to bind c in closure
-                next_buf = self._iopool.submit((lambda x: (lambda: image.read(tiles[x + 1][0])))(c))
-            if cur_buf is None:
-                continue
+        epoch = self._epoch[1 if is_labels else 0]
```
Comment (ScottMcMichael) [inline delta/imagery/imagery_dataset.py:125]: Add comments to the new blocks of code here summarizing what is going on

File: delta/imagery/imagery_dataset.py [hunk 3]
```diff
@@ -253,42 +121,66 @@ def _tile_generator(self, i, is_labels): # pragma: no cover
...
+        image_locks = {}
+        image_preprocesses = {}
+        for img in images:
+            image_locks[img] = threading.Lock()
+            image_preprocesses[img] = img.get_preprocess()
+            img.set_preprocess(None) # parallelize preprocessing outside lock
+
+        def tile_gen():
+            # use same seed for labels and not labels, differ by epoch
+            rand = random.Random(self._random_seed + epoch * 11617)
```
Comment (ScottMcMichael) [inline delta/imagery/imagery_dataset.py:137]: This is just a large prime number?

File: delta/imagery/imagery_dataset.py [hunk 4]
```diff
@@ -253,42 +121,66 @@ def _tile_generator(self, i, is_labels): # pragma: no cover
...
+        gen = tile_gen()
+        buf_queue = []
+        for _ in range(config.io.threads() * 2): # add a bit ahead
+            try:
+                next_item = next(gen)
+            except StopIteration:
+                break
+            add_to_queue(buf_queue, next_item)
+        while buf_queue:
+            (img, (_, sub_tiles), buf) = buf_queue.pop(0)
```
Comment (ScottMcMichael) [inline delta/imagery/imagery_dataset.py:174]: Why is just one of these outputs "blanked out" while only 'buf' is used?

File: delta/imagery/imagery_dataset.py (hunk 3 above)
Comment (bcoltin) [reply delta/imagery/imagery_dataset.py:137]: yes

File: delta/imagery/imagery_dataset.py (hunk 4 above)
Comment (bcoltin) [reply delta/imagery/imagery_dataset.py:174]: no, good point

**Review by ScottMcMichael — APPROVED** [review]

**Pull request merged**